modules/ducktests/tests/ignitetest/tests/dns_failure_test.py

```diff
@@ -0,0 +1,134 @@
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Module contains DNS service failure test.
+"""
+import os
+import socket
+
+from ignitetest.services.ignite import IgniteService
+from ignitetest.services.ignite_app import IgniteApplicationService
+from ignitetest.services.utils.control_utility import ControlUtility
+from ignitetest.services.utils.ignite_aware import IgniteAwareService
+from ignitetest.services.utils.ignite_configuration import IgniteConfiguration, DataStorageConfiguration
+from ignitetest.services.utils.ignite_configuration.data_storage import DataRegionConfiguration
+from ignitetest.services.utils.ignite_configuration.discovery import from_ignite_cluster
+from ignitetest.utils import ignite_versions
+from ignitetest.utils.ignite_test import IgniteTest
+from ignitetest.utils.version import IgniteVersion, DEV_BRANCH
+from ignitetest.utils import cluster
+
+
+class DnsFailureTest(IgniteTest):
+    """
+    Test DNS service failure.
+    """
+
+    @cluster(num_nodes=4)
+    @ignite_versions(str(DEV_BRANCH))
+    def dns_failure_test(self, ignite_version):
+        """
+        DNS failure test.
+        """
+        # Replace hosts with IP addresses.
+        for node in self.test_context.cluster.nodes:
+            node.account.externally_routable_ip = socket.gethostbyname(node.account.externally_routable_ip)
+
+        version = IgniteVersion(ignite_version)
+
+        ignite_config = IgniteConfiguration(
+            version=version,
+            data_storage=DataStorageConfiguration(
+                default=DataRegionConfiguration(
+                    persistence_enabled=True
+                )
+            )
+        )
+
+        ignites = self.__prepare_service(ignite_config, 3)
+
+        self.__unblock_dns(ignites)
+
+        # Start nodes one-by-one to reproduce the problem.
+        ignites.start_node(ignites.nodes[0])
+        ignites.await_started([ignites.nodes[0]])
+        ignites.start_node(ignites.nodes[1])
+        ignites.await_started([ignites.nodes[0], ignites.nodes[1]])
+        ignites.start_node(ignites.nodes[2])
+        ignites.await_started()
+
+        control_utility = ControlUtility(ignites)
+        control_utility.activate()
+
+        self.__block_dns(ignites, 20000, True)
+
+        ignites.stop_node(ignites.nodes[1])
+
+        ignites.await_event("Node left topology", 60, from_the_beginning=True,
+                            nodes=[ignites.nodes[0], ignites.nodes[2]])
+
+        assert ignites.alive(ignites.nodes[0]), 'Node 0 should be alive'
+        assert ignites.alive(ignites.nodes[2]), 'Node 2 should be alive'
+
+        self.__unblock_dns(ignites, [ignites.nodes[1]])
+
+        ignites.start_node(ignites.nodes[1])
```

Comment:
What if node starts with blocked DNS? Do we have test for it? Check the stacktrace:

`getHostByAddr:-1, Inet6AddressImpl (java.net)
getHostByAddr:936, InetAddress$PlatformNameService (java.net)
getHostFromNameService:659, InetAddress (java.net)
getHostName:602, InetAddress (java.net)
getHostName:574, InetAddress (java.net)
addresses:2343, IgniteUtils (org.apache.ignite.internal.util)
resolveLocalAddresses:2302, IgniteUtils (org.apache.ignite.internal.util)
resolveLocalAddresses:2265, IgniteUtils (org.apache.ignite.internal.util)
getNodeAttributes:843, TcpCommunicationConfigInitializer (org.apache.ignite.spi.communication.tcp.internal)
parseNodeAttributes:734, GridManagerAdapter (org.apache.ignite.internal.managers)
startSpi:285, GridManagerAdapter (org.apache.ignite.internal.managers)
start:487, GridIoManager (org.apache.ignite.internal.managers.communication)
startManager:1772, IgniteKernal (org.apache.ignite.internal)
start:1007, IgniteKernal (org.apache.ignite.internal)
start0:1725, IgnitionEx$IgniteNamedInstance (org.apache.ignite.internal)
start:1647, IgnitionEx$IgniteNamedInstance (org.apache.ignite.internal)
start0:1089, IgnitionEx (org.apache.ignite.internal)
start:599, IgnitionEx (org.apache.ignite.internal)`

What if `getHostByAddr:-1, Inet6AddressImpl (java.net)` is blocked ?

Reply:
In this case node start time is much longer (timeouts should be fixed), but overall, solution is still workable.
I don't think we need to fix resolveLocalAddresses and resolveLocalHost, since main goal of this ticket is to prevent node failures with minimal inconvenience. Fixing these methods is risky, since local host name can be used as consistentId in some cases, so, we should do our best to resolve this hostname properly.
Also, there are some calls to resolve hostname inside third-party libraries (log4j, spring as far as I remember), which we can't avoid.